alexkarpitski/cloud-custodian
# bazel - test coverage

**Bazel test coverage rule**

To run tests with coverage, make sure that:
- your target or its base folder is added into [.bazel-coveragerc](https://github.com/alexkarpitski/cloud-custodian/pull/26/files#diff-6bed5a1d20d2d9d8a6464249fb2085d7)
- your target or its dependencies has  `//:.bazel-coveragerc` [in data](https://github.com/alexkarpitski/cloud-custodian/pull/26/files#diff-b8d3bcb05bd672235f0f7abd610e9c62)

Examples of usage:
`bazel coverage //tests`
`bazel coverage //tests:[test_name]`
`bazel coverage //tools/c7n_gcp/tests`
`bazel coverage //tools/c7n_azure/tests_azure`
`bazel coverage //tools/c7n_kube/tests`
`bazel coverage //tools/c7n_mailer/tests`

To see the report:
1. Install **coverage.py**: `pip install coverage`
2. Open coverage folder in terminal: `cd /tmp/C7N-cov/[module.name]`
3. Use _coverage.py_ commands to see the report:
   - `coverage combine` - combines coverage data files;
   - `coverage report` - terminal report;
   - `coverage html` - html report;
4. Useful options:
   `--skip-empty` - to exclude empty files from report;
   `--include "[regular expression]"` - to include files matching the regular expression. For example, `coverage report --include "*/c7n_gcp/*"` to include files only from _c7n_gcp_ module.

See docs: [Coverage.py command line usage](https://coverage.readthedocs.io/en/coverage-5.0.4/cmd.html)


The rule uses **[coverage.py](https://coverage.readthedocs.io/en/coverage-5.0.4/)** tool to retrieve coverage data. 

Environment variable **COVERAGE_FILE** is used to specify the coverage data file path.

**.bazel-coveragerc** - coverage configuration file used to specify sources to cover. Also parallel mode is set here, since bazel runs tests in parallel. It allows to keep the data separate during measurement, later coverage files can be combined into a single **.coverage** file with `coverage combine` command.

**Coverage.py** requires test files to have permanent location, so files from sandbox cannot be used. That's why **/home/user/.cache/bazel/\_bazel\_user/hash of the workspace dir/execroot/\_\_main\_\_** directory is used. See more: [Output directory layout](https://docs.bazel.build/versions/master/output_directories.html)
